### ISSUE TITLE
Question: Support for Slack threads?

### ISSUE THREAD

**Issue by coltenkrauter** [GregHilston/Simple-Slack-Bot#35]
How would I go about having the Slack bot respond in a Slack thread? I am familiar with the Slack API, but I'm not seeing a simple way to respond to a message in a thread. 

Is that something that this simple-slack-bot framework is capable of doing? 

EDIT: I modified that title to include reactions as in the comments section I have started a conversation about adding support for threads and reactions.

**Comment by GregHilston** [comment]
I don't believe I ever worked with threads. What I would do is have your
bot listen in a channel for messages and send a thread message in that
channel.

Look at the payload that the Slack bot receives and feel free to share it
here.

I can see if we can make a few minor changes or improvements to handle
threads

**Comment by coltenkrauter** [comment]
Okay thanks, I'll take a look 👍

**Comment by coltenkrauter** [comment]
So, I may or may not get to this, but I am thinking that this codebase would become far more versatile if we could add support for threading and reactions (emojis).

My initial thoughts are that we could add thread support to the **write** function in [this file](https://github.com/GregHilston/Simple-Slack-Bot/blob/master/simple_slack_bot/slack_request.py#L56). The end state, perhaps there is an optional function parameter for choosing to respond in the given thread. A boolean switch type of thing. 

I am thinking that we could add a new function called **react** which would be very similar to the **write** function, but rather than posting messages in channels or threads, it would simply set (add or remove) Slack emojis to given messages.

I am looking into this further. 

Basically, I want a library that supports this type of functionality which is very simple, such as the one you've written. I'm leaning towards contributing to your codebase rather than writing my own... because why recreate the wheel. 

What are your thoughts @GregHilston?

**Comment by GregHilston** [comment]
@coltenkrauter let me start off by addressing your last part first:

> I'm leaning towards contributing to your codebase rather than writing my own... because why recreate the wheel.

I am 100% down with you pull requesting any and all changes to this library if you want to avoid recreating the wheel. I can't promise you how much time I'll spend actively developing on this but I will look over every and all PR and provide help when I can.

> My initial thoughts are that we could add thread support to the write function in this file. The end state, perhaps there is an optional function parameter for choosing to respond in the given thread. A boolean switch type of thing.

Sounds perfect to me.

> I am thinking that we could add a new function called react which would be very similar to the write function, but rather than posting messages in channels or threads, it would simply set (add or remove) Slack emojis to given messages.

Also sounds perfect.

Let me see if I can get the project to run on my end and get back to you on these two topics above

**Comment by GregHilston** [comment]
@coltenkrauter I got the project running just fine on my end. I also merged all those pull requests AND remembered how to push the package up to pypi and did that. 

The newest version is `1.3.2` and has been officially uploaded. Just so we don't forget, I put the commands in the `Makefile`.

So I think everything is a go to continue development!

**Comment by GregHilston** [comment]
BTW: Looked quickly into threads. it looks like a threaded message gets passed a `thread_ts` that links to the timestamp of the parent.

Parent:

```
received an event of type message and slack event {'client_msg_id': '6712d61e-4b18-4ad8-893e-3a481a5fb83a', 'suppress_notification': False, 'type': 'message', 'text': 'This is a test on a thread', 'user': 'U389
XUHT2', 'team': 'T38AP97Q9', 'blocks': [{'type': 'rich_text', 'block_id': 'UdV', 'elements': [{'type': 'rich_text_section', 'elements': [{'type': 'text', 'text': 'This is a test on a thread'}]}]}], 'thread_ts':
 '1585943493.003900', 'source_team': 'T38AP97Q9', 'user_team': 'T38AP97Q9', 'channel': 'C3FAX3UCV', 'event_ts': '1585943499.004000', 'ts': '1585943499.004000'}
received an event of type message and slack event {'type': 'message', 'subtype': 'message_replied', 'hidden': True, 'message': {'client_msg_id': '685721b0-affb-4b06-b480-85622cdd4ad9', 'type': 'message', 'text'
: 'This is a test of the base message', 'user': 'U389XUHT2', 'ts': '1585943493.003900', 'team': 'T38AP97Q9', 'blocks': [{'type': 'rich_text', 'block_id': '2/c', 'elements': [{'type': 'rich_text_section', 'eleme
nts': [{'type': 'text', 'text': 'This is a test of the base message'}]}]}], 'thread_ts': '1585943493.003900', 'reply_count': 2, 'reply_users_count': 1, 'latest_reply': '1585943501.004200', 'reply_users': ['U389
XUHT2']}, 'channel': 'C3FAX3UCV', 'event_ts': '1585943501.004300', 'ts': '1585943501.004300'}
```

thread message
```
received an event of type message and slack event {'client_msg_id': '4261a6b5-72c8-456d-a7c6-ebcef24b3efc', 'suppress_notification': False, 'type': 'message', 'text': 'ping', 'user': 'U389XUHT2', 'team': 'T38AP
97Q9', 'blocks': [{'type': 'rich_text', 'block_id': 'G=XwL', 'elements': [{'type': 'rich_text_section', 'elements': [{'type': 'text', 'text': 'ping'}]}]}], 'thread_ts': '1585943493.003900', 'source_team': 'T38A
P97Q9', 'user_team': 'T38AP97Q9', 'channel': 'C3FAX3UCV', 'event_ts': '1585943501.004200', 'ts': '1585943501.004200'}
```

Since simple slack bot uses [slacker](https://github.com/os/slacker) under the hood to interact with the Slack API, I quickly looked for how to reply to a thread and couldn't find a quick solution. I'll have to look later but please read above and let me know your thoughts

**Comment by coltenkrauter** [comment]
Great! I intend to contribute, but it may be a week or two before I submit a PR.

**Comment by GregHilston** [comment]
Take all the time you need! If you'd like to have more conversations feel free to reach out to me at 

Gregory.Hilston@gmail.com

and we can keep some of the planning off of this Github issue. Your call. 

Happy to be of any assistance

**Comment by coltenkrauter** [comment]
I'm open to that. Thanks! I'll touch base soon.

**Comment by coltenkrauter** [comment]
Update: I've been pretty busy... Not sure when I can get to this, but it is still my intention to contribute at some point.

**Comment by GregHilston** [comment]
Hey not a problem at all! Whenever you find time, feel free to reach out to me here and we can chat a bit about ways to approach.

No rush, just happy you're even still thinking about this project

**Comment by GregHilston** [comment]
@coltenkrauter Not poking you for any response on your part, but I'll be using this framework for a small side project in the near future. My plan is to discuss with you what I've learned and see what you think. I'll keep ya posted

**Comment by coltenkrauter** [comment]
Awesome 👍

I have been busy one some other projects so I don't have anything for you yet.

**Comment by GregHilston** [comment]
@coltenkrauter

No worries at all mate! And it isn't for me but for the open source project :)

If I get some inspiration, I'll look into the request this weekend :) would love to have more people using it

**Comment by GregHilston** [comment]
@coltenkrauter Alright great news, was able to get this project to reply to channels this morning.

The current code is not pretty, so I'm going to spend a bit of time cleaning it up before I post a PR. I'd love if you reviewed the PR before we merge it.

I'm shelving reactions for now, just for time reasons, but can look at it in the future. 

I'll bother you when a PR is posted

**Comment by coltenkrauter** [comment]
Sounds good, I'll review the PR 👍

**Comment by GregHilston** [comment]
Since most of this dialogue revolved around threads, I'm going to make a new one just for reactions to keep it easier for me to follow.
